crates/bevy_reflect/src/generics.rs

```diff
@@ -0,0 +1,338 @@
+use crate::type_info::impl_type_methods;
+use crate::{Reflect, Type, TypePath};
+use alloc::borrow::Cow;
+use alloc::sync::Arc;
+use core::ops::Deref;
+
+/// The generic parameters of a type.
+///
+/// This is automatically generated via the [`Reflect` derive macro]
+/// and stored on the [`TypeInfo`] returned by [`Typed::type_info`]
+/// for types that have generics.
+///
+/// It supports both type parameters and const parameters
+/// so long as they implement [`TypePath`].
+///
+/// If the type has no generics, this will be empty.
+///
+/// If the type is marked with `#[reflect(type_path = false)]`,
+/// the generics will be empty even if the type has generics.
+///
+/// [`Reflect` derive macro]: bevy_reflect_derive::Reflect
+/// [`TypeInfo`]: crate::type_info::TypeInfo
+/// [`Typed::type_info`]: crate::Typed::type_info
+#[derive(Clone, Default, Debug)]
+pub struct Generics(Box<[GenericInfo]>);
+
+impl Generics {
+    /// Creates an empty set of generics.
+    pub fn new() -> Self {
+        Self(Box::new([]))
+    }
+
+    /// Finds the generic parameter with the given name.
+    ///
+    /// Returns `None` if no such parameter exists.
+    pub fn get_named(&self, name: &str) -> Option<&GenericInfo> {
+        // For small sets of generics (the most common case),
+        // a linear search is often faster using a `HashMap`.
+        self.0.iter().find(|info| info.name() == name)
+    }
+
+    /// Adds the given generic parameter to the set.
+    pub fn with(mut self, info: impl Into<GenericInfo>) -> Self {
+        self.0 = IntoIterator::into_iter(self.0)
+            .chain(core::iter::once(info.into()))
+            .collect();
+        self
+    }
+}
+
+impl<T: Into<GenericInfo>> FromIterator<T> for Generics {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        Self(iter.into_iter().map(Into::into).collect())
+    }
+}
+
+impl Deref for Generics {
+    type Target = [GenericInfo];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// An enum representing a generic parameter.
+#[derive(Clone, Debug)]
+pub enum GenericInfo {
+    /// A type parameter.
+    ///
+    /// An example would be `T` in `struct Foo<T, U>`.
+    Type(TypeParamInfo),
+    /// A const parameter.
+    ///
+    /// An example would be `N` in `struct Foo<const N: usize>`.
+    Const(ConstParamInfo),
+}
+
+impl GenericInfo {
```

Comment:
Question: Do we have a way to know trait bounds of a generic? If so, is it worth to store?

Reply:
I considered it but I’m not sure it's really worth storing. Could be interesting to explore in a future PR if the need for it arises (and the apparent limitations are acceptable).